plugins/trip.hpp

```diff
@@ -0,0 +1,332 @@
+                                 }
+                                 return true;
+                             }(),
+                             "invalid component ranges");
+          };
+
+        // constructor to use when whole graph is one single scc
+        SCC_Component(std::vector<NodeID> in_component)
+                      : component(in_component), range({0, in_component.size()}) {
+                      };
+
+        std::size_t GetNumberOfComponents() const{
+            return range.size() - 1;
+        }
+
+        const std::vector<NodeID> component;
+        // component range = in_range + [component.size()]
+        std::vector<size_t> range;
+    };
+
+    // takes the number of locations and its distance matrix,
+    // identifies and splits the graph in its strongly connected components (scc)
+    // and returns an SCC_Component
+    SCC_Component SplitUnaccessibleLocations(const std::size_t number_of_locations,
+                                             const DistTableWrapper<EdgeWeight> & result_table) {
+
+        // Run TarjanSCC
+        auto wrapper = std::make_shared<MatrixGraphWrapper<EdgeWeight>>(result_table.GetTable(), number_of_locations);
+        auto scc = TarjanSCC<MatrixGraphWrapper<EdgeWeight>>(wrapper);
+        scc.run();
+
+        std::vector<size_t> range_insertion;
+        std::vector<size_t> range;
+        range_insertion.reserve(scc.get_number_of_components());
+        range.reserve(scc.get_number_of_components());
+
+        std::vector<NodeID> components(number_of_locations, 0);
+
+        auto prefix = 0;
+        for (size_t j = 0; j < scc.get_number_of_components(); ++j){
+            range_insertion.push_back(prefix);
+            range.push_back(prefix);
+            prefix += scc.get_component_size(j);
+        }
+
+        for (size_t i = 0; i < number_of_locations; ++i) {
+            components[range_insertion[scc.get_component_id(i)]] = i;
+            ++range_insertion[scc.get_component_id(i)];
+        }
+
+        return SCC_Component(components, range);
+    }
+
+    void SetLocPermutationOutput(const std::vector<NodeID> & permutation,
+                                 osrm::json::Object & json_result){
+        osrm::json::Array json_permutation;
+        json_permutation.values.insert(std::end(json_permutation.values),
+                                       std::begin(permutation),
+                                       std::end(permutation));
+        json_result.values["permutation"] = json_permutation;
+    }
+
+    void ComputeRoute(const PhantomNodeArray & phantom_node_vector,
+                      const RouteParameters & route_parameters,
+                      const std::vector<NodeID> & trip,
+                      InternalRouteResult & min_route) {
+        // given he final trip, compute total distance and return the route and location permutation
+        PhantomNodes viapoint;
+        for (auto it = std::begin(trip); it != std::end(trip); ++it) {
```

Comment:
This calls `std::end(trip)` on every iteration. Improve this by storing the end iterator (because it does not change):

    for (auto it = std::begin(trip), end = std::end(trip); it != end; ++it)

Or even better, more idiomatic and describes intend (and it really cool):

Let's say you have a trip `1 -> 2 -> 3 -> 4 -> 1`. What you really want is to connect all elements with their following one and the last one to the first one. This can be done by taking the `trip` vector, rotating it to the left by one, then zip-ing both ranges with a function that generates the `viapoint` and outputs it into the `segment_end_coordinates`.

    trip:        1  2  3  4
                  \  \  \  \
    out:           +  +  +  +
                  /  /  /  /
    trip_rot:    2  3  4  1

Here are the two key lines you need for this:

    std::rotate(begin(trip_rot), begin(trip_rot) + 1, end(trip_rot));
    std::transform(begin(trip), end(trip), begin(trip_rot), out, toViapoint);

where `out` is e.g. a `std::back_inserter` applied to the `segment_end_coordinates` like:

    auto out = std::back_inserter(min_route.segment_end_coordinates)

and `toViapoint` is a lambda that takes to `NodeId`s and returns a `PhantomNodes` containing the vias, like:

    auto toViapoint = [&] (const NodeId& lhs, const NodeId& rhs) {
        return PhantomNodes{phantom_node_vector[lhs][0], phantom_node_vector[rhs][0]};
    };

Hope that helps.

Refs.
 - http://en.cppreference.com/w/cpp/algorithm/rotate
 - http://en.cppreference.com/w/cpp/algorithm/transform